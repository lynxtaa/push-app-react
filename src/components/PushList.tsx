import React from 'react'

interface Props {
	children: React.ReactNode
}

const PushList = ({ children }: Props) => (
	<ul className="pushlist list-unstyled">{children}</ul>
)

export default PushList
