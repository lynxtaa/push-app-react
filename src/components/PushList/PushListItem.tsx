import { useSpring, animated, config } from 'react-spring'
import { Button, ListItem } from '@chakra-ui/react'

interface Props {
	children: React.ReactNode
	isHidden?: boolean
	onClick(event: React.MouseEvent<HTMLButtonElement, MouseEvent>): void
}

const AnimatedListItem = animated(ListItem)

export default function PushListItem({ onClick, children, isHidden }: Props) {
	const spring = useSpring({
		from: { opacity: 0, marginBottom: '0px', height: '0px' },
		to: { opacity: 1, marginBottom: '16px', height: '40px' },
		config: config.gentle,
		reverse: isHidden,
	})

	return (
		<AnimatedListItem
			style={{
				...spring,
				visibility: (spring as any).opacity.interpolate((o: number) =>
					o === 0 ? 'hidden' : 'visible',
				),
			}}
		>
			<Button variant="outline" w="100%" onClick={onClick} colorScheme="brand">
				{children}
			</Button>
		</AnimatedListItem>
	)
}
