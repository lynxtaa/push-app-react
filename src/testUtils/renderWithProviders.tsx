import { render, RenderOptions, RenderResult } from '@testing-library/react'
import { ChakraProvider } from '@chakra-ui/react'
import { MemoryRouter } from 'react-router-dom'

import theme from '../theme'

export default function renderWithProviders(
	ui: React.ReactElement,
	{
		initialEntries,
		initialIndex,
		...rest
	}: Omit<RenderOptions, 'queries'> & {
		initialEntries?: string[]
		initialIndex?: number
	} = {},
): RenderResult {
	return render(ui, {
		...rest,
		wrapper: ({ children }: { children?: React.ReactNode }) => (
			<ChakraProvider theme={theme}>
				<MemoryRouter initialEntries={initialEntries} initialIndex={initialIndex}>
					{children}
				</MemoryRouter>
			</ChakraProvider>
		),
	})
}
