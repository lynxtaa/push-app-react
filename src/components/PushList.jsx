import React from 'react'
import PropTypes from 'prop-types'

import PushListItem from './PushListItem'

const PushList = ({ sets }) => (
	<ul className="pushlist">
		{ sets.map((set, index) => <PushListItem set={set} key={index} />) }
	</ul>
)

PushList.propTypes = {
	sets: PropTypes.arrayOf(PropTypes.number).isRequired,
}

export default PushList
