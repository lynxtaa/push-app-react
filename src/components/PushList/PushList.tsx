import { List } from '@chakra-ui/core'
import React, { useState, useEffect } from 'react'

import PushListItem from './PushListItem'

interface Set {
	id: string
	set: number
}

type Props = {
	sets: Set[]
}

export default function PushList({ sets }: Props) {
	const [hidden, setHidden] = useState<string[]>([])

	useEffect(() => {
		setHidden([])
	}, [sets])

	return (
		<List>
			{sets.map(({ set, id }) => (
				<PushListItem
					key={id}
					onClick={() => setHidden([...hidden, id])}
					isHidden={hidden.includes(id)}
				>
					{set}
				</PushListItem>
			))}
		</List>
	)
}
