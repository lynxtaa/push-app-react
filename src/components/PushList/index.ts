import PushList from './PushList'

export default PushList
