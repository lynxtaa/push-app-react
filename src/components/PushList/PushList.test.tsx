import React from 'react'
import { screen } from '@testing-library/react'
import userEvent from '@testing-library/user-event'
import createMockRaf, { MockRaf } from '@react-spring/mock-raf'

import PushList from './PushList'
import renderWithProviders from 'testUtils/renderWithProviders'

let mockRaf: MockRaf

beforeEach(() => {
	mockRaf = createMockRaf()
	window.requestAnimationFrame = mockRaf.raf
})

const sets = [
	{ id: '1', set: 1 },
	{ id: '2', set: 2 },
	{ id: '3', set: 3 },
]

it('renders list', () => {
	const { container } = renderWithProviders(<PushList sets={sets} />)
	expect(container.firstChild).toMatchSnapshot()
})

it('hides item after click', async () => {
	renderWithProviders(<PushList sets={sets} />)

	const item = screen.getByText('2')
	userEvent.click(item)
	mockRaf.flush()

	expect(item).not.toBeVisible()
})
