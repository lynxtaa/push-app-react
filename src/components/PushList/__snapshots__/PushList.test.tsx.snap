// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders list 1`] = `
<ul
  class="css-155za0w"
  role="list"
>
  <li
    class="css-0"
    style="opacity: 0; margin-bottom: 0px; height: 0px; visibility: hidden;"
  >
    <button
      class="chakra-button css-q6bbzb"
      type="button"
    >
      1
    </button>
  </li>
  <li
    class="css-0"
    style="opacity: 0; margin-bottom: 0px; height: 0px; visibility: hidden;"
  >
    <button
      class="chakra-button css-q6bbzb"
      type="button"
    >
      2
    </button>
  </li>
  <li
    class="css-0"
    style="opacity: 0; margin-bottom: 0px; height: 0px; visibility: hidden;"
  >
    <button
      class="chakra-button css-q6bbzb"
      type="button"
    >
      3
    </button>
  </li>
</ul>
`;
